_posts/2019-07-10-feature-columns/feature_columns.Rmd

```diff
@@ -0,0 +1,566 @@
+
+model <- keras_model(inputs, output)
+```
+
+Now, if you've actually read through the whole of this part, you may wish for an easier way to get to this point. So let's switch to feature specs for the rest of this post.
+
+## Feature specs to the rescue
+
+In spirit, the way feature specs are defined follows the example of the  [recipes package](https://cran.r-project.org/web/packages/recipes/index.html). (It won't make you hungry, though.)  You initialize a feature spec with the prediction target -- `feature_spec(target ~ .)`, and then use the `%>%` to tell it what to do with individual columns. "What to do" here signifies two things:
+
+- First, how to "read in" the data. Are they numeric or categorical, and if categorical, what am I supposed to do with them? For example, should I treat all distinct symbols as distinct, resulting in, potentially, an enormous count of categories -- or should I constrain myself to a fixed number of entities? Or hash them, even? 
+- Second, optional subsequent transformations. Numeric columns may be bucketized; categorical columns may be embedded. Or features could be combined to capture interaction.
+
+In this post, we demonstrate the use of a subset of `step_` functions. The vignettes on [Feature columns](https://tensorflow.rstudio.com/tools/tfdatasets/articles/feature_columns.html) and [Feature specs](https://tensorflow.rstudio.com/tools/tfdatasets/articles/feature_spec.html) illustrate additional functions and their application. 
+
+Starting from the beginning again, here is the complete code for data read-in and train-test split in the feature spec version. 
+
+```{r}
+library(keras)
+library(tfdatasets)
+library(readr)
+library(dplyr)
+
+path <- "train.csv"
+
+porto <- read_csv(path)
+
+porto <- porto %>%
+  mutate_at(vars(ends_with("cat")), as.character)
+
+id_training <- sample.int(nrow(porto), size = 0.75*nrow(porto))
+training <- porto[id_training,]
+testing <- porto[-id_training,]
+```
+
+Data-prep-wise, recall what our goals are: leave alone if binary; scale if numeric; embed if categorical.
+Specifying all of this does not need more than a few lines of code:
+
+```{r}
+ft_spec <- training %>%
+  select(-id) %>%
+  feature_spec(target ~ .) %>%
+  step_numeric_column(ends_with("bin")) %>%
+  step_numeric_column(-ends_with("bin"),
+                      -ends_with("cat"),
+                      normalizer_fn = scaler_standard()
+                      ) %>%
+  step_categorical_column_with_vocabulary_list(ends_with("cat")) %>%
+  step_embedding_column(ends_with("cat"),
+                        dimension = function(vocab_size) as.integer(sqrt(vocab_size) + 1)
+                        ) %>%
+  fit()
+```
+
+
+Note how here we are passing in the training set, and just like with `recipes`, we won't need to repeat any of the steps for the validation set. Scaling is taken care of by `scaler_standard()`, an optional transformation function passed in to `step_numeric_column`.
```

Comment:
Is there a function reference for `scaler_standard()` somewhere? If so, link to it here!